packages/shell-api/src/decorators.ts

```diff
@@ -34,6 +50,55 @@ export class ShellApiClass implements ShellApiInterface {
   }
 }
 
+// For classes like Collection, it can be useful to attach information to the
+// result about the original data source, so that downstream consumers of the
+// shell can e.g. figure out how to edit a document returned from the shell.
+// To that end, we wrap the methods of a class, and report back how the
+// result was generated.
+// We also attach the `shellApiType` and `asShellResult` properties to the
+// return type (if that is possible and they are not already present), so that
+// we can also provide sensible information for methods that do not return
+// shell classes, like db.coll.findOne() which returns a Document (i.e. a plain
+// JavaScript object).
+function wrapWithAddSourceToResult(fn: Function, functionName: string): Function {
+  function addSource<T extends {}>(result: T, obj: any, args: any[]): T {
+    if (typeof result === 'object' && result !== null) {
+      const resultSourceInformation: ShellResultSourceInformation = {
+        dataSource: obj._dataSourceIdentifier(),
+        call: functionName,
```

Comment:
Yeah, I’ve gone back and forth on this a bit myself … it feels like something that could be really useful, e.g. for figuring out how to get the data that this call resulted in a second time or displaying what call resulted in this output. I’d lean towards keeping it since it’s essentially free and data that is always going to be available anyway, but I’m also happy to remove it